kernel/src/loader/mod.rs

```diff
@@ -258,6 +264,87 @@ pub fn load_cmdline(
     Ok(())
 }
 
+/// Loads the initrd from a file into the given memory slice.
+///
+/// * `guest_mem` - The guest memory the initrd is written to.
+/// * `initrd_image` - Input initrd image.
+///
+/// Returns the entry address of the initrd and its length as a tuple
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+pub fn load_initrd<F>(
+    guest_mem: &GuestMemory,
+    initrd_image: &mut F,
+) -> Result<(GuestAddress, usize)>
+where
+    F: Read + Seek,
+{
+    let initrd_size: usize;
+    match initrd_image.seek(SeekFrom::End(0)) {
+        Ok(0) | Err(..) => return Err(Error::ReadInitrd),
+        Ok(s) => initrd_size = s as usize,
+    };
+
+    let align_to_pagesize = |address| address & !(PAGE_SIZE - 1);
+
+    initrd_image
+        .seek(SeekFrom::Start(0))
+        .map_err(|_| Error::ReadInitrd)?;
+
+    let region_size: usize = guest_mem.region_size(0).map_err(|_| Error::LoadInitrd)?;
```

Comment:
What is the reason for loading the initrd at the end of the first memory region? I may be missing something but in order to get the end of the low memory you should do: 
```
guest_mem.end_addr()
```

Reply:
You are right, I need the "end" of the low memory. I though `end_addr()` would return addressed above 4G.

Reply:
@dianpopa, actually here I need the size of the low memory region, not the end address, since I am assuming the region may not start at offset 0.